### DISCUSSION TITLE
What is the correct way to delete a submission after calling an integration?

### DISCUSSION THREAD

**Discussion by roland-d** [verbb/formie#1135] [Q&A] [answered]
Hello,

The scenario I am looking at is that when an integration is called successfully we want to delete the submission. This is all working except for the fact that after the deletion Formie wants to go back to the submission but it no longer exists.

In my integration after the payload has been delivered and all is successful, I call `Craft::$app->getElements()->deleteElement($submission, true);` to delete the submission. This successfully deletes the submission. 

Where it goes wrong is when I am in the GUI doing the following:
1. I open a submission
2. Select the integration from the dropdown
3. The integration runs successfully and calls Craft to delete the submission
4. Craft deletes the submission
5. Formie wants to load the submission in `src/controllers/SubmissionsController.php` but fails on line 174 with this code because the submission is empty.
```
if (!$variables['submission']) {
            throw new HttpException(404);
        }
```

Am I calling the deletion method wrong?

**Comment by engram-design** [comment]
Right, so if you're triggering the integration when editing the submission from the control panel, you're definitely going to get a 404 error when trying to load the edit page of the submission again - because it will be gone! Your delete code looks fine to me, and that should work correctly.

What would you expect to have happen? To be redirected to the Submissions index after running your integration?

If so, that would be tricky. There's no guarantee that the integration will be run immediately (it's run via the queue after all) and you'd need to wait for that to be run before the submission is deleted. But in your case, it must be running immediately.

I'm hesitant to add handling in that controller to redirect back to the submission index if the submission can't be found, because that's strange behaviour for regular use-cases. If I try and visit a submission in the control panel and it keeps redirecting me back to the submissions index, that might seem like a error.

**Comment by roland-d** [comment]
Hello @engram-design I understand your point of view regarding the redirect. Not redirecting to the list view is fine for me.

As for `that should work correctly` I am running into another issue. The job is not processed correctly because it comes back with this error:
```
session_set_cookie_params(): Session cookie parameters cannot be changed after headers have already been sent
```
Removing the `deleteElement` line does make the job end successfully. Can you have a look at that?

**Reply by engram-design** [reply]
Are you able to send through the full stack trace of the error? Not _quite_ sure where that error is coming from in the scheme of things.

And if you can post an excerpt of your custom integration class showing where you delete the submission, just so I'm replicating your setup.

**Comment by roland-d** [comment]
@engram-design 

Here is the full stack trace of the error:
```
08:47:12 88128 #0 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/web/ErrorHandler.php(79): yii\base\ErrorHandler->handleError(2, 'session_set_coo...', '/var/www/html/b...', 430)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'session_set_coo...', '/var/www/html/b...', 430)
#2 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/web/Session.php(430): session_set_cookie_params(Array)
#3 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/web/Session.php(148): yii\web\Session->setCookieParamsInternal()
#4 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/web/Session.php(723): yii\web\Session->open()
#5 /var/www/html/DOMAIN/server/vendor/verbb/formie/src/elements/Form.php(1940): yii\web\Session->remove('formie:87897:sn...')
#6 /var/www/html/DOMAIN/server/vendor/verbb/formie/src/elements/Form.php(968): verbb\formie\elements\Form->resetSnapshotData()
#7 /var/www/html/DOMAIN/server/vendor/verbb/formie/src/elements/Submission.php(1075): verbb\formie\elements\Form->getCurrentSubmission()
#8 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/services/Elements.php(1709): verbb\formie\elements\Submission->beforeDelete()
#9 /var/www/html/DOMAIN/server/plugins/greenhouse/src/modules/Greenhouse.php(282): craft\services\Elements->deleteElement(Object(verbb\formie\elements\Submission), true)
#10 /var/www/html/DOMAIN/server/vendor/verbb/formie/src/services/Submissions.php(283): perfectwebteam\greenhouse\modules\Greenhouse->sendPayload(Object(verbb\formie\elements\Submission))
#11 /var/www/html/DOMAIN/server/vendor/verbb/formie/src/jobs/TriggerIntegration.php(63): verbb\formie\services\Submissions->sendIntegrationPayload(Object(perfectwebteam\greenhouse\modules\Greenhouse), Object(verbb\formie\elements\Submission))
#12 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2-queue/src/Queue.php(246): verbb\formie\jobs\TriggerIntegration->execute(Object(craft\queue\Queue))
#13 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2-queue/src/cli/Queue.php(147): yii\queue\Queue->handleMessage(383620, 'O:36:"verbb\\for...', 300, 1)
#14 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/queue/Queue.php(182): yii\queue\cli\Queue->handleMessage(383620, 'O:36:"verbb\\for...', 300, 1)
#15 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/queue/Queue.php(157): craft\queue\Queue->executeJob()
#16 [internal function]: craft\queue\Queue->craft\queue\{closure}(Object(Closure))
#17 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2-queue/src/cli/Queue.php(117): call_user_func(Object(Closure), Object(Closure))
#18 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/queue/Queue.php(165): yii\queue\cli\Queue->runWorker(Object(Closure))
#19 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/controllers/QueueController.php(82): craft\queue\Queue->run()
#20 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/controllers/QueueController.php(103): craft\controllers\QueueController->actionRun()
#21 [internal function]: craft\controllers\QueueController->actionRetry()
#22 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#23 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#24 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction('retry', Array)
#25 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/web/Application.php(301): yii\base\Module->runAction('queue/retry', Array)
#26 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/web/Application.php(625): craft\web\Application->runAction('queue/retry', Array)
#27 /var/www/html/DOMAIN/server/vendor/craftcms/cms/src/web/Application.php(280): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#28 /var/www/html/DOMAIN/server/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application->handleRequest(Object(craft\web\Request))
#29 /var/www/html/DOMAIN/server/public_html/index.php(26): yii\base\Application->run()
#30 {main}
```
The excerpt from my custom integration:
```php
public function sendPayload(Submission $submission): bool
    {
        try {
           
            $payload = [];

            $response = $this->deliverPayload($submission, 'jobs/' . $formValues['id'], $payload, 'POST', 'multipart');

            Greenhouse::log($submission->id . ' ' . json_encode($response));

            // This line fails with the session error
            Craft::$app->getElements()->deleteElement($submission, true);
        } catch (Exception $exception) {
            \perfectwebteam\greenhouse\Greenhouse::log($submission->id . ' ' . $exception->getTraceAsString());
            Integration::error($this, $exception->getMessage(), true);
        }

        return true;
    }
```

What might be good to know is that this is a form with 2 File Upload fields and the form is set to delete the files when the submission is deleted.

Let me know if this is enough.

**Reply by engram-design** [reply]
Ah right, so yes that makes some sense. How are you running your queue? This seems to be the error output when the integration is run via the queue, which is having issues with the session-removal.

**Comment by roland-d** [comment]
I have actually tried both, via the GUI and via CLI. From what I understand the following happens:

- I submit the form
- GraphQL sends it to Formie
- Formie processes the form and stores the submission
- The submission is added to the queue

When I then visit the queue via the GUI, I see it failed with the session error. I also tried not to visit the GUI and run `./craft queue/run` but this also ends with the session error.

**Reply by engram-design** [reply]
So running the queue via the CLI shouldn't throw any session issues, because we have guards setup to determine if it's a console command. The queue by default will run so long as you have a page open in the control panel.

Just wanted to check your `runQueueAutomatically` general.php setting is disabled when trying to run from the CLI?

I've also been trying both, but unable to replicate it. But I can see the lines in the trace causing the issue (just seemingly not an issue for me):

```
#5 /var/www/html/backbase.com/server/vendor/verbb/formie/src/elements/Form.php(1940): yii\web\Session->remove('formie:87897:sn...')
#6 /var/www/html/backbase.com/server/vendor/verbb/formie/src/elements/Form.php(968): verbb\formie\elements\Form->resetSnapshotData()
#7 /var/www/html/backbase.com/server/vendor/verbb/formie/src/elements/Submission.php(1075): verbb\formie\elements\Form->getCurrentSubmission()
#8 /var/www/html/backbase.com/server/vendor/craftcms/cms/src/services/Elements.php(1709): verbb\formie\elements\Submission->beforeDelete()
```

Bottom-up, running `beforeDelete()` on the submission, which triggers `resetSnapshotData()` which checks against a console request (in this case it isn't) and has an issue with the session.

**Comment by roland-d** [comment]
> The queue by default will run so long as you have a page open in the control panel.

That I was not aware of. So I redid my test, only have the CLI open. Here is what I did:

- I submitted a form
- `./craft queue/info` To see the job
- `./craft queue/run` To process the job
- Job was processed without a problem

This is my `general.php` setting for running the queue:
`'runQueueAutomatically' => true,`

I have set this to `false` for my CLI test.

As you said, this won't be a problem running it via CLI. Just somewhat problematic via the GUI because the payload has already been sent in that case and we do not want to send it again.

**Reply by engram-design** [reply, marked as answer]
Fixed for the next release. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.13",
  "...": "..."
}
```

Then run `composer update`.
